### PR TITLE
Upgrade debian to bookworm - release on 2023-06-10

### PR DESCRIPTION
**Pull request by ieugen**
https://wiki.debian.org/DebianBookworm

Debian bookworm is releasing in ~ 10 days. 
Code wise there will be no breaking changes, only bug fixes.

I believe it's time to consider publishing an updated version based on bookworm.
For our use case we need a newer release of buildah. 

### Testing done

make build
make test

Will push and use in our infra soon. 

### Submitter checklist

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by dduportal** [comment]
Hi @ieugen , thanks for this contribution!

Until the publication date, we have to choices:

- We can either wait for the 10 June before merging this PR, as default agent version must always be on stable/LTS operating systems
- Or you can update your PR to add the `bookworm` along with `bullseye` in the BuildX bake file, keeping the `stable` tag to `bullseye` to benefit sooner from this PR

**Comment by ieugen** [comment]
Thanks for the quick reply. 
I will add the tag. 
I think it's good to have some version overlap (provided it's not too hard on maintenance). 
Maintain bullseye for a few month / year. 
Have the next version of Debian as an image to encourage testing + disclaimer that it might be buggy.

**Comment by dduportal** [comment]
> Thanks for the quick reply. I will add the tag. I think it's good to have some version overlap (provided it's not too hard on maintenance). Maintain bullseye for a few month / year. Have the next version of Debian as an image to encourage testing + disclaimer that it might be buggy.

I agree with you, this is a good idea to maintain an overlap to let users migrate without breaking changes.
Let' s see how easy it is on code basis first with the Bake file :)

**Comment by ieugen** [comment]
@dduportal : PR updated.

File: README.md
Comment (dduportal) [inline README.md:83]: Can you add JDK11 as well?

File: debian/Dockerfile
Comment (dduportal) [inline debian/Dockerfile:2]: I (just now while reviewing) realize that this change as an impact: Dependabot won't update the version anymore (as it only supports the `FROM` instruction).

We have to switch to `updatecli` instead (as we already have some update manifests in the `updatecli/` directory, executed by a GitHub Action)

File: docker-bake.hcl
Comment (dduportal) [inline docker-bake.hcl:60]: Same comment as for the Dependabot update: we have to track this version

File: docker-bake.hcl
Comment (dduportal) [inline docker-bake.hcl:139]:
```suggestion
    equal(ON_TAG, "true") ? "${REGISTRY}/${JENKINS_REPO}:bookworm-${VERSION}-jdk17": "",
```

**Review by dduportal — CHANGES_REQUESTED** [review]
> @dduportal : PR updated. Bookworm supports only jdk17. There is no trixie docker image support yet https://hub.docker.com/_/debian . I think it will pop up sometime this year, a few months after the release. Once that is provided, we can update the build.

Thanks for this update! It is really a good idea to introduce a `next` version.

A few elements to note, blocking the PR (alas :'():

- Do you mind adding the JDK11 version please? As we use the Eclipse temuring JDK distribution (instead of the Debian `openjdk` package), there should be no obstacle to do this
- There is a tag to be updated in the bakefile (`VERSION` is the version number of the image when built from a tag) to avoid overriding the current version once merged
- We have to keep the automatic update of the base image: Either tricking Dependabot to continue updating the parent image version(s) (e.g. using Dockerfiles with `FROM` instructions) or add an `updatecli` manifest to update the 2 variables you've added in the bakefile

File: README.md
Comment (ieugen) [reply README.md:83]: sure.

File: docker-bake.hcl
Comment (dduportal) [reply docker-bake.hcl:60]: No problem, I can help if you want to split the burden?

File: README.md
Comment (ieugen) [reply README.md:83]: Done.

File: docker-bake.hcl
Comment (ieugen) [reply docker-bake.hcl:65]: I don't mind. 
I would rather see this merged faster than slower :) .

File: debian/Dockerfile
Comment (ieugen) [reply debian/Dockerfile:2]: Relates to https://github.com/jenkinsci/docker-ssh-agent/issues/242 .

File: docker-bake.hcl
Comment (ieugen) [reply docker-bake.hcl:65]: I think only the automatic version update is left to do.

File: debian/Dockerfile
Comment (gounthar) [reply debian/Dockerfile:2]: > I (just now while reviewing) realize that this change as an impact: Dependabot won't update the version anymore (as it only supports the `FROM` instruction).
> 
> We have to switch to `updatecli` instead (as we already have some update manifests in the `updatecli/` directory, executed by a GitHub Action)

How to track the Debian release? I haven't found (yet) a GitHub repo that would list the Debian releases.
Can `updatecli` parse an HTML file once retrieved via `http`?

Comment (dduportal) [reply debian/Dockerfile:2]: > How to track the Debian release?  I haven't found (yet) a GitHub repo that would list the Debian releases.
Can updatecli parse an HTML file once retrieved via http?

I guess you want to track the Docker image's version (like Dependabot does today). In that case, the https://www.updatecli.io/docs/plugins/resource/docker_image/ source should be enough.

Given the proposal from @ieugen, we should start by pinning the versions of both Bullseye and Bookworm in the PR here, and then do a subsequent PR with updatecli.

**Comment by dduportal** [comment]
Hello @ieugen , this PR looks good and is almost ready to roll.

The context changed since my last comments:

- @poddingue started a huge work on the dependency updates with `updatecli` and his method shows it is easier to start by pinning dependencies, and then send a subsequent PR for tracking the pinned dependency.
- The bookworm distribution has been officially released.


As such, WDYT about:
- Rebasing this PR against the latest version of the principal branch (a release 5.5.0 was made earlier today)
- Ensure that we pinn both Debian versions and do not worry about Dependabot or `updatecli` (we will work on this after your PR is shipped)

?

File: docker-bake.hcl
Comment (ieugen) [reply docker-bake.hcl:65]: Let's solve this in the new PR.

**Comment by ieugen** [comment]
Duplicate of #262

**Pull request closed without merging**
